File: Array/3143. Maximum Points Inside the Square.cpp
/*
3143. Maximum Points Inside the Square
*/

// TC:O(nlgn) SC:O(n)
class Solution {
public:
    int maxPointsInsideSquare(vector<vector<int>>& points, string s) {
        int n = points.size();
       
        vector<pair<int,int>> arr;
        for(int i = 0; i < n; i++){
            int a = abs(points[i][0]),  b = abs(points[i][1]);
            int take = max(a, b);
            arr.push_back({take, i});
        }
        sort(arr.begin(), arr.end());
        
        int res = INT_MIN;
        
        int square = 0;
        int idx = 0;
        int count = 0;
        unordered_map<int,int> map;
        while(idx < arr.size()){
            bool flag = 0;
            square = max(square, arr[idx].first);
            
            while(idx < n && square >= arr[idx].first){
                int ch = s[arr[idx].second]-'a';
                map[ch] += 1;
                count += 1;
                idx++;
                if(map[ch] >= 2) {
                    flag = 1;
                    break;
                }
            }
            
            if(flag) break;
            
           res = max(res, count);
            
            
        }
        
        return res == INT_MIN ? 0 : res;
        
    }
};

/*
從0 開始出發的點 包起來的正方形 不能包含兩個相同的字母 問最多可以取到多少點
包起來的定義是邊邊跟內含都算
去計算每個點到原點距離取最大的那個邊長 從小排到大 開始判斷
*/

File: Array/3143. Maximum Points Inside the Square.py
#3143. Maximum Points Inside the Square


# TC:O(nlgn) SC:O(n)
class Solution:
    def maxPointsInsideSquare(self, points: List[List[int]], s: str) -> int:
        dict_of_set = defaultdict(set) # 可以創建一個哈希表 沒有賦值的會自動value 為 set()

        for (x,y),c in zip(points, s):
            c = ord(c) - ord('a') # 將字元轉成0-25, ord 為Unicode 
            d = max(abs(x), abs(y))
            if c in dict_of_set[d]:
                dict_of_set[d].add(-1)
            else :
                dict_of_set[d].add(c)
        
        union = {-1}
        for d, st in sorted(dict_of_set.items()): # sorted 函式 
            if st & union:
                break
            union |= st
        return len(union)-1
